add meson.build to integrate libbpf as a subproject of a project built by Meson

wat-ze-hex opened the pull request:
Make it possible to include libbpf into projects built by Meson build system as
a subproject. This is the case e.g. for systemd.

meson.build declares dependencies which may be used in parent projects:
`libbpf_static_dep` for static library
`libbpf_shared_dep` for shared library
`bpf_includes` provides the location of bpf.h, btf.h and libbpf.h

Parent projects need to introduce git wrap to build libbpf:
https://github.com/mesonbuild/meson/blob/master/docs/markdown/Wrap-dependency-system-manual.md#wrap-git

An alternative approach is applying a patch with meson.build for libbpf while
building a parent project:
https://github.com/mesonbuild/meson/blob/master/docs/markdown/Wrap-dependency-system-manual.md#wrap-file
Imo it's less transparent and would add more headache.

Note that this commit provides only the essentials for adding libbpf as a
subproject, it does not install headers or libraries.

meson.build is placed in the source root because that's how Meson expects it to
be placed.

The related discussion is in https://github.com/systemd/systemd/pull/12151

evverx commented on meson.build:
```diff
@@ -0,0 +1,77 @@
+        join_paths(libbpf_src_dir, 'src/str_error.c'),
+        join_paths(libbpf_src_dir, 'src/str_error.h'),
+        join_paths(libbpf_src_dir, 'src/xsk.c'),
+        join_paths(libbpf_src_dir, 'src/xsk.h'))
```
@rdna would it be possible to set up Travis CI or Semaphore CI that would run `meson ./build && ninja -C ./build` to make sure that `libbpf` could be built with `meson`? If there were any tests it could also run `ninja -C ./build test` to be 100% sure that nothing was forgotten.

yonghong-song replied:
I am not familiar with Travis CI or Semaphore CI. But I think it is a good idea to integrate them if possible to add test coverage. But that is a different subject than this pull request.

evverx replied:
This PR is supposed to make it possible to build `libbpf` with `meson` and it seems reasonable to me to add something that would help to make sure that it works and it won't be broken in a week because it's easy to forget to handle this file. It just seems to me that the systemd CI isn't the right place to make sure that `libbpf` can be built with `clang` and `-werror` for example.

evverx replied:
I just realized that it's possible to point `meson` to @wat-ze-hex's repository in https://github.com/systemd/systemd/pull/12151 to get around the issue with submodules on Ubuntu CI (where I just turned on ARM just in case) (that is, there is no need to wait for this PR to be merged). It seems to be easier and faster at this point.

rdna replied:
@evverx yeah, that's actually very good idea. We can do it either separately, as @yonghong-song suggested, or here, no strong preference.

I'm not familiar with these CI either, but trying to play with Semaphore now to set it up for my fork of libbpf. I see it has many different settings in `semaphore.yml` so if you have any recommendations on how to set it up to keep testing procedure as close as possible to systemd process that'd be really helpful. E.g. I see that by default it uses `ubuntu1804` image -- not sure if it's good enough for our first step.  We can start with something simple and extend it along the way.

wat-ze-hex replied:
> Is there a way to use some masks here (`*.h` / `*.c` ?) like it's done in `src/Makefile`? So that it doesn't go out of sync when new files are added to the library since it's easy to forget to handle this file.

Can't see that Meson has regex builtin. Well, as a last resort we can shell out:
```
libbpf_sources = files(run_command('find',
        ['@0@/src'.format(libbpf_src_dir), '-type',  'f', '-name', '*.h',  '-o',  '-name', '*.c']).stdout().split())
```
Looks ugly but it works.

@evverx
>I just realized that it's possible to point meson to @wat-ze-hex's repository in systemd/systemd#12151 to get around the issue with submodules on Ubuntu CI (where I just turned on ARM just in case) (that is, there is no need to wait for this PR to be merged). It seems to be easier and faster at this point.
I had this idea but then there was another one: that pointing to my fork tastes like anarchy :) I can do that.
Is there way to set up Ubuntu CI just like Travis CI or Semaphore CI ?

evverx replied:
@rdna judging by `ubuntu1804`, it seems to be Semaphore 2.0, which I've never used unfortunately. We use "classic" Semaphore, where we run debian-testing containers on top of Ubuntu 16.04, which is probably overkill given that you already have access to a more or less fresh build environment. I'll try to take a look at the documentation to figure out how to install `gcc`, `clang`, `libelf`, `meson` and `ninja` and run `meson && ninja` there.

@wat-ze-hex I agree pointing `meson` to your repository would have been a bit chaotic (in a good way) but it would help to keep experimenting with `libbpf` in that PR without going through the libbpf repository where you need to think about the licence headers and other boring and time-consuming stuff. That I think could be done later (including the CI part on the libbpf side).

Regarding Ubuntu CI, it was set up by the debian and ubuntu maintainers a long time ago. I'm not saying it isn' possible but I'm afraid I can't be of any help here.

rdna replied:
@evverx thank you for all the details! Just to double-check .. do you recommend to go ahead with Semaphore 2.0 and this `ubuntu1804` image? It seems this process may take some time due to learning curve so any help on this is appreciated :)

evverx replied:
@rdna I'm looking at the "pricing" page there and it seems to me that Semaphore 2.0 isn't exactly free for open-source project (though, they accept all the credit cards, which is reassuring :-)). I'll try to dig deeper later but right now I can't say I'd choose it.

In the meantime, I took a couple of scripts from `systemd` and changed them so that they could build `libbpf` only (with `gcc`, `clang` and also with ASAN/UBsan but I guess that stages can be safely removed because they don't make much sense without tests). Judging by https://github.com/evverx/libbpf/pull/1, it's more or less working and in theory if Travis CI was turned on, it would be possible to add [the second commit](https://github.com/evverx/libbpf/pull/1/commits/fff6f1df402f23e23c3c5541fa648abd020dad85) (or its variation) here to get the ball rolling.

wat-ze-hex replied:
> I just realized that it's possible to point `meson` to @wat-ze-hex's repository in [systemd/systemd#12151](https://github.com/systemd/systemd/pull/12151) to get around the issue with submodules on Ubuntu CI (where I just turned on ARM just in case) (that is, there is no need to wait for this PR to be merged). It seems to be easier and faster at this point.

Done, still getting failures. The problem is for sure that it's explicitly forbidden to download wraps in Ubuntu CI, they have `--wrap-mode=nodownload`
https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-bionic-upstream-systemd-ci-systemd-ci/bionic/arm64/s/systemd-upstream/20190411_035033_c0833@/log.gz

So, from this point it's only fixing the tests.

Also LGTM failed. Think due to `--recurse-submodules=no` is on, but I'm not sure.

evverx replied:
> The problem is for sure that it's explicitly forbidden to download wraps in Ubuntu CI, they have --wrap-mode=nodownload

@wat-ze-hex I didn't know that. That means that it doesn't seem to be possible to do anything without the maintainers of Ubuntu CI (and I already CCed them in that PR). So it appears the switch wasn't necessary. I'm sorry about that.

Regrading LGTM, to judge from https://discuss.lgtm.com/t/lgtm-seems-to-be-failing-to-extract-the-systemd-project-due-to-a-submodule/1994/4?u=evverx they use some kind of cache that should be updated manually but I guess it would be better to wait until everything else is fixed before asking them to update the cache. Plus I turned LGTM on in https://github.com/evverx/libbpf/pull/1 and it didn't find anything so the alerts are the same: https://lgtm.com/projects/g/evverx/libbpf/alerts/?mode=tree.

I'm not sure why the tests are still failing under UBSan though. In https://github.com/libbpf/libbpf/issues/28 I applied the patch and the issue wasn't there anymore. Is that patch included in the branch that was tested under UBsan?

wat-ze-hex replied:
>So it appears the switch wasn't necessary. I'm sorry about that.

No problem, that's the right approach so libbpf maintainers can keep `meson.build` in sync with Makefile. Also if meson provides capabilities to manage subprojects, we should utilize them.

> I'm not sure why the tests are still failing under UBSan though. In #28 I applied the patch and the issue wasn't there anymore. Is that patch included in the branch that was tested under UBsan?

My `master` doesn't have the fix. Waiting for https://patchwork.ozlabs.org/patch/1083690/ in libbpf upstream.

rdna commented on meson.build:
```diff
@@ -0,0 +1,68 @@
+        include_directories : libbpf_includes,
+        link_depends : libbpf_map,
+        install : false,
+        c_args : libbpf_c_args)
```
I see from documentation that `libbpf_shared` supports `version` and `soversion` and uses reasonable values by default, but the result it provides with this configuration differs from what `Makefile` does.

I built libbpf both with meson in `build/` directory and with `Makefile` in `src/`.

`SONAME` should be fixed:
```
% diff -u <(readelf -d src/libbpf.so  | grep SONAME) <(readelf -d build/libbpf.so  | grep SONAME)
--- /dev/fd/63  2019-04-10 12:57:06.593038040 -0700
+++ /dev/fd/62  2019-04-10 12:57:06.594038036 -0700
@@ -1 +1 @@
- 0x000000000000000e (SONAME)             Library soname: [libbpf.so.0]
+ 0x000000000000000e (SONAME)             Library soname: [libbpf.so]
```

libelf should be linked:
```
% diff -u <(ldd src/libbpf.so | cut -d" " -f 3 | grep lib) <(ldd build/libbpf.so | cut -d" " -f 3 | grep lib)
--- /dev/fd/63  2019-04-10 13:03:56.108390558 -0700
+++ /dev/fd/62  2019-04-10 13:03:56.109390554 -0700
@@ -1,3 +1 @@
-/lib64/libelf.so.1
 /lib64/libc.so.6
-/lib64/libz.so.1
```

Versioning script `libbpf.map` should be used (not it seems it's not):
```
diff -u <(readelf -s --wide src/libbpf.so | grep -Eo '[^ ]+@LIBBPF_.*' | cut -d@ -f 3 | sort | uniq -c) <(readelf -s --wide build/libbpf.so | grep -Eo '[^ ]+@LIBBPF_.*' | cut -d@ -f 3 | sort | uniq -c)
--- /dev/fd/63  2019-04-10 13:05:29.687021896 -0700
+++ /dev/fd/62  2019-04-10 13:05:29.688021892 -0700
@@ -1,2 +0,0 @@
-    121 LIBBPF_0.0.1
-     29 LIBBPF_0.0.2
```

Also I"m not sure how specifically a project will use shared lib, but it might make sense to provide version symlinks same way `Makefile` does it now:
```
% ls -al src/libbpf.so*
lrwxrwxrwx. 1 rdna users     11 Apr 10 12:55 src/libbpf.so -> libbpf.so.0
lrwxrwxrwx. 1 rdna users     15 Apr 10 12:55 src/libbpf.so.0 -> libbpf.so.0.0.2
-rwxr-xr-x. 1 rdna users 377632 Apr 10 12:55 src/libbpf.so.0.0.2
```

See `Makefile` for details on how it does all these things.

bluca replied:
To find the libelf dependency in a way that's compatible with both a system with pkg-config and without, this snippet can be used:

```
libelf = dependency('libelf', required: false)
if not libelf.found()
        libelf = cc.find_library('elf', required: true)
endif

...

shared_library(...
    dependencies: libelf,
...
```

wat-ze-hex replied:
@bluca It looks Meson handles this internally on `master` https://github.com/mesonbuild/meson/blob/master/mesonbuild/dependencies/base.py#L572-L592

I'm a bit surprised that Meson doesn't handle this internally, studying the code just in case if it does.

wat-ze-hex replied:
@rdna 
Discrepancies you mentioned are gone: https://github.com/libbpf/libbpf/commit/86ed4a6cfdc41886d8cff2f7dd120566759f92bb

Also` install` part is similar to `make install` with the difference that there is no check for `BUILD_STATIC_ONLY` envar, so both `*.a` and `*.so` are installed. It doesn't matter for systemd use case.

bluca replied:
It doesn't fall back automatically unfortunately and it's mightily annoying, there's an issue open on their repo that asks for the feature but it's not done yet

wat-ze-hex commented:
@yonghong-song 
I was able to set up Travis CI for my fork and got undefined-behavior repro with asan: https://travis-ci.org/wat-ze-hex/systemd
This can be done in a few clicks, could you please check it?  /* an access to the libbpf account in required */

yonghong-song commented:
@wat-ze-hex We fixed a few libbpf bugs in last several days. I will do a sync from upstream, and you can run again. There could be more minor things to fix which I did not pick them up.

yonghong-song commented:
@wat-ze-hex could you try again? I just pulled in latest upstream change and let us see whether we still have blocking errors (undefined behavior, potential segfault, etc.) from Travis CI  or not.

evverx commented:
@yonghong-song judging by https://travis-ci.org/evverx/libbpf/jobs/518532279 libbpf is failing to compile now:
```
[11/12] cc -Ibpf@sta -I. -I.. -I.././include -I.././include/uapi -fdiagnostics-color=always -fsanitize=address,undefined -fno-omit-frame-pointer -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -g -fPIC -g -O2 -Werror -Wall -Wno-pointer-arith -Wno-sign-compare  -MD -MQ 'bpf@sta/src_libbpf.c.o' -MF 'bpf@sta/src_libbpf.c.o.d' -o 'bpf@sta/src_libbpf.c.o' -c ../src/libbpf.c
FAILED: bpf@sta/src_libbpf.c.o 
cc -Ibpf@sta -I. -I.. -I.././include -I.././include/uapi -fdiagnostics-color=always -fsanitize=address,undefined -fno-omit-frame-pointer -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -g -fPIC -g -O2 -Werror -Wall -Wno-pointer-arith -Wno-sign-compare  -MD -MQ 'bpf@sta/src_libbpf.c.o' -MF 'bpf@sta/src_libbpf.c.o.d' -o 'bpf@sta/src_libbpf.c.o' -c ../src/libbpf.c
../src/libbpf.c: In function 'bpf_object__elf_collect':
../src/libbpf.c:947:18: error: 'map_def_sz' may be used uninitialized in this function [-Werror=maybe-uninitialized]
   if (map_def_sz <= sizeof(struct bpf_map_def)) {
       ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/libbpf.c:827:18: note: 'map_def_sz' was declared here
  int i, map_idx, map_def_sz, nr_syms, nr_maps = 0, nr_maps_glob = 0;
                  ^~~~~~~~~~
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

yonghong-song commented:
The code is introduced by Daniel B's static variable support.
```
commit d859900c4c56dc4f0f8894c92a01dad86917453e
Author: Daniel Borkmann <daniel@iogearbox.net>
Date:   Tue Apr 9 23:20:13 2019 +0200

    bpf, libbpf: support global data/bss/rodata sections
    
    This work adds BPF loader support for global data sections
    to libbpf. This allows to write BPF programs in more natural
    C-like way by being able to define global variables and const
    data.
...
```

For the following code,
```
        if (obj->efile.maps_shndx >= 0) {
                Elf_Scn *scn = elf_getscn(obj->efile.elf,
                                          obj->efile.maps_shndx);

                if (scn)
                        data = elf_getdata(scn, NULL);
                if (!scn || !data) {
                        pr_warning("failed to get Elf_Data from map section %d\n",
                                   obj->efile.maps_shndx);
                        return -EINVAL;
                }
        }
```
The tools thinks obj->efile.maps_shndx may be negative which probably not sure (need to double check). The bpf program needs maps to be useful.

@rdna could you help submit a patch to bpf-next? just giving map_def_sz an init value 0 to silence the error.

yonghong-song commented:
This shows travis CI does help! We should bring it in to libbpf repo.

rdna commented:
@yonghong-song, @evverx patch is out: https://patchwork.ozlabs.org/patch/1083690/

yonghong-song commented:
The Andrey's patch which fixed new libbpf Travis CI error (https://patchwork.ozlabs.org/patch/1083690/) just get merged into libbpf.

evverx commented:
Travis CI in https://github.com/evverx/libbpf/pull/1 is green meaning that libbpf no longer fails to compile. Although I'm pretty sure the unit tests under UBsan should be fine, I'm wondering if the kernel panic is gone. Originally, I thought it was caused by UBsan too but having taken a closer look, it seems that systemd crashed in VMs and didn't in containers, which looks very suspicious because that code is triggered everywhere. @wat-ze-hex could you temporarily switch to the submodules approach to get around those "AttributeError: 'NoneType' object has no attribute 'startswith" failures I see at https://travis-ci.org/systemd/systemd/jobs/518646441.

wat-ze-hex commented:
> @wat-ze-hex could you temporarily switch to the submodules approach to get around those "AttributeError: 'NoneType' object has no attribute 'startswith" failures I see at https://travis-ci.org/systemd/systemd/jobs/518646441 to make sure that the kernel panic is gone?

Looks like `NoneType` was me messing with meson. I've just synced my master with upstream and asan tests are still failing:
https://travis-ci.org/systemd/systemd/builds/518923927?utm_source=github_status&utm_medium=notification

evverx commented:
That was unexpected :-) I'll try to reproduce it manually. Just out of curiosity, did you change any compiler flags like `-O`, for example?

evverx commented:
I've just updated the submodule linked to that PR and it seems that it is out of date:
```c
        attr.map_flags = create_attr->map_flags;
        memcpy(attr.map_name, create_attr->name,
               min(name_len, BPF_OBJ_NAME_LEN - 1));
        attr.numa_node = create_attr->numa_node;
```

wat-ze-hex commented:
> I've just updated the submodule linked to that PR and it seems that it is out of date:
> 
> ```c
>         attr.map_flags = create_attr->map_flags;
>         memcpy(attr.map_name, create_attr->name,
>                min(name_len, BPF_OBJ_NAME_LEN - 1));
>         attr.numa_node = create_attr->numa_node;
> ```

Hmm, but how?
These are commits on my `master`:
https://github.com/wat-ze-hex/libbpf/commits/master 
They include https://github.com/wat-ze-hex/libbpf/commit/1770ac49cf2f722db64d992b2f201002b8c6cfc0

That PR linked to 
https://github.com/systemd/systemd/pull/12151/commits/f8e925389e08b7d95856d61150b2c3a0676ba0f6#diff-1a31bf68c329d0ce599517aac94f4cb4R3

evverx commented:
The submodule refers to https://github.com/wat-ze-hex/libbpf/commit/e335e90a0bcbdf1689da112347c04a76d37ed4df and it isn't based on the last version of libbpf. As far as I can tell, in that PR you use both the submodule and the subproject (which refers to the head of your repository). Looks like the submodule takes precedence over the subproject.

evverx commented:
@wat-ze-hex I've updated the submodule locally and added the following commit on top of your branch:
```

    subprojects/libbpf

diff --git a/subprojects/libbpf b/subprojects/libbpf
index e335e90a0b..3b8723d83d 160000
--- a/subprojects/libbpf
+++ b/subprojects/libbpf
@@ -1 +1 @@
-Subproject commit e335e90a0bcbdf1689da112347c04a76d37ed4df
+Subproject commit 3b8723d83d79b2db5f582e6934a949e58e667095
```
and everything seems to be fine in the sense that the tests passed. But I have to admit I'm new to the concept of "wraps" so I'm not sure whether it's expected or not.

wat-ze-hex commented:
> @wat-ze-hex I've updated the submodule locally and added the following commit on top of your branch:

Thanks, updated the PR.
What I did:
```
git fetch wat-ze-hex.libbpf.origin
git reset --hard wat-ze-hex.libbpf.origin/master
```
Guess this crosses out the whole submodule approach. We can us this for testing while the RPM is not out, but since the `Subproject commit` has to be updated by systemd maintainers, that's a no go.

wat-ze-hex commented:
> Guess this crosses out the whole submodule approach. We can us this for testing while the RPM is not out, but since the `Subproject commit` has to be updated by systemd maintainers, that's a no go.

On the second thought this approach introduces more flexible versioning than the versioning of RPM package which is tied to ABI.
BTW, Travis is green :)

evverx commented:
I agree that, in a way, submodules are more flexible than regular packaged libraries but at this point I'm not sure it's safe to rely on the source code that doesn't seem to be tested regularly. I mean, I'm not sure that I can simply update the submodule so that it can be built successfully. Plus, if the code is included in the repository it will go through the testing machinery including https://scan.coverity.com/ (where we send the source code every day and which is a bit more sophisticated than `-Werror`). Personally I'm not sure I'm ready to go through issues it will find in `libbpf` and report them here for example. But don't get me wrong, I think `libbpf` should be used in `systemd` and it would be great to keep experimenting with it in the PR you opened to see what the code is going to look like when you replace micro-assembly with C. I also hope ASan won't complain about putting "compiled" code into systemd, reading them and loading them into the kernel :-)

> BTW, Travis is green :)

I'm glad the kernel panic is gone :-) I suspect It didn't crash in containers due to UBsan because judging by "Not running with unified cgroups, BPF firewalling is not supported." systemd inherited the legacy cgroup hierarchy from the host and bailed out before it was able to call `bpf_create_map`. In VMs it used the default hybrid hierarchy, called `bpf_create_map` and failed.

bluca commented on meson.build:
```diff
@@ -0,0 +1,68 @@
+# SPDX-License-Identifier: LGPL-2.1 OR BSD-2-Clause
+project('libbpf', 'c',
+        version : '0.0.2',
```
there's another PR that is changing this to 0.0.3 (as it is in the kernel)

wat-ze-hex replied:
Resolved, thanks!

bluca commented:
Could you please add the configuration to generate the pkg-config file?
Should be something like:

```
pkg = import('pkgconfig')
pkg.generate(name: meson.project_name(),
        version: meson.project_version(),
        libraries: libbpf_shared,
        libraries_private: libbpf_static,
        description: '''BPF library'''
)
```

Thanks!

wat-ze-hex commented:
@bluca

Updated PR. Also this is consistent with `Makefile` with the exception that `*.pc` generated by `make` doesn't have 
`Libs.private: -L${libdir} -lbpf` field:

```
diff src/libbpf.pc /usr/lib64/pkgconfig/libbpf.pc
--- src/libbpf.pc	2019-04-12 16:20:04.185629738 -0700
+++ /usr/lib64/pkgconfig/libbpf.pc	2019-04-12 19:13:17.397609440 -0700
@@ -1,12 +1,11 @@
-# SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
-
 prefix=/usr
-libdir=/usr/lib64
+libdir=${prefix}/lib64
 includedir=${prefix}/include

 Name: libbpf
 Description: BPF library
 Version: 0.0.3
-Libs: -L${libdir} -lbpf
 Requires.private: libelf
+Libs: -L${libdir} -lbpf
+Libs.private: -L${libdir} -lbpf
 Cflags: -I${includedir}
```

bluca commented:
The libs.private as generated there is redundant, so it can be ignored, thanks

bluca commented:
Ah I might know why it's added: the libraries_static field I suggested might be redundant, could you please try to remove it and see if it works? The requires.private should still be generated

rdna commented on meson.build:
```diff
@@ -0,0 +1,91 @@
+        link_args : libbpf_link_args,
+        link_depends : libbpf_map_source_path,
+        soversion : patchlevel,
+        version : extraversion)
```
@wat-ze-hex symlinks are still created differently:
```
diff -u <(cd src && stat -c %N libbpf.so* | sort) <(cd build && stat -c %N libbpf.so* | sort)
--- /dev/fd/63  2019-04-15 10:20:53.029672412 -0700
+++ /dev/fd/62  2019-04-15 10:20:53.030672409 -0700
@@ -1,3 +1,3 @@
-‘libbpf.so.0.0.3’
-‘libbpf.so.0’ -> ‘libbpf.so.0.0.3’
+‘libbpf.so.0’ -> ‘libbpf.so.3’
+‘libbpf.so.3’
 ‘libbpf.so’ -> ‘libbpf.so.0’
```

See [corresponding code](https://github.com/libbpf/libbpf/blob/8bbb21c227d64efe3ad96cfd427984e51f6ee549/src/Makefile#L73-L82) in `src/Makefile`.

I assume only `extraversion` is used where the whole version should be.

wat-ze-hex replied:
My bad. Fixed:
```
$ diff -u <(cd src && stat -c %N libbpf.so* | sort) <(cd build && stat -c %N libbpf.so* | sort)
$
```

rdna commented on meson.build:
```diff
@@ -0,0 +1,91 @@
+        )
+
+patchlevel = '0'
+extraversion = '3'
```
Nit: Let's reduce number of places that should be kept in sync. Version components can be specified once and then used in `project.version` to make full version, [same way as in `src/Makefile`](https://github.com/libbpf/libbpf/blob/8bbb21c227d64efe3ad96cfd427984e51f6ee549/src/Makefile#L3-L7).

wat-ze-hex replied:
@rdna 

Thanks, fixed. Splitting `version` string instead of joining due to
`ERROR: First statement must be a call to project`
Duplications are gone.

wat-ze-hex commented:
> Ah I might know why it's added: the libraries_static field I suggested might be redundant, could you please try to remove it and see if it works? The requires.private should still be generated

Removed `Libs.private`:

```
diff src/libbpf.pc /usr/lib64/pkgconfig/libbpf.pc
--- src/libbpf.pc	2019-04-15 14:02:31.493198251 -0700
+++ /usr/lib64/pkgconfig/libbpf.pc	2019-04-15 14:25:48.634165432 -0700
@@ -1,12 +1,10 @@
-# SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
-
 prefix=/usr
-libdir=/usr/lib64
+libdir=${prefix}/lib64
 includedir=${prefix}/include

 Name: libbpf
 Description: BPF library
 Version: 0.0.3
-Libs: -L${libdir} -lbpf
 Requires.private: libelf
+Libs: -L${libdir} -lbpf
 Cflags: -I${includedir}
```

The only thing is missing is `License` field but I don't see Meson has a keyword for that: https://mesonbuild.com/Pkgconfig-module.html
